src/definitions/x509/crl.rs

```diff
@@ -0,0 +1,320 @@
+    SignatureTypeMismatch(Box<AlgorithmIdentifierOwned>, Box<AlgorithmIdentifierOwned>),
+    #[error("Unable to extract CRL sequence signature")]
+    ExtractCrlSequenceSignature(asn1_rs::Err<asn1_rs::Error>),
+    #[error("Missing curve")]
+    MissingCurve,
+    #[error("Unknown curve extension: {0}")]
+    UnknownCurveExt(ObjectIdentifier),
+    #[error("Missing public key")]
+    MissingPublicKey,
+    #[error("Signature check of CRL failed")]
+    SignatureCheckOfCrlFailed,
+    #[error("Invalid public key format")]
+    InvalidPublicKeyFormat,
+    #[error("Missing signature")]
+    MissingSignature,
+    #[error("Signature in wrong format")]
+    SignatureInWrongFormat,
+}
+
+pub const OID_EC_CURVE_P256: ObjectIdentifier = ObjectIdentifier::new_unwrap("1.2.840.10045.3.1.7");
+pub const OID_EC_CURVE_P384: ObjectIdentifier = ObjectIdentifier::new_unwrap("1.3.132.0.34");
+pub const OID_PUBLIC_KEY_ELLIPTIC_CURVE: ObjectIdentifier =
+    ObjectIdentifier::new_unwrap("1.2.840.10045.2.1");
+
+const OID_EXTENSION_REASON_CODE: ObjectIdentifier = ObjectIdentifier::new_unwrap("2.5.29.21");
+
+/// Given a cert, download and verify the associated crl listed in the cert, and verify the crl
+/// against the cert's metadata and public key.
+pub async fn fetch_and_validate_crl(root_cert: &TbsCertificate) -> Result<Vec<TbsCertList>, Error> {
+    let distribution_points = match read_distribution_points(root_cert)? {
+        None => return Ok(vec![]),
+        Some(distribution_points) => distribution_points,
+    };
+
+    let mut cert_lists = vec![];
+
+    for distribution_point in distribution_points.0.iter() {
+        let distribution_point_name = distribution_point.distribution_point.as_ref().ok_or(
+            Error::DistributionPointMalformed("missing distributionPoint name"),
+        )?;
+
+        let urls = distribution_point_urls(distribution_point_name)?;
+
+        for url in urls.iter() {
+            let crl_bytes = fetch_crl(url).await?;
+
+            cert_lists.push(validate_crl(root_cert, &crl_bytes)?);
+        }
+    }
+
+    Ok(cert_lists)
+}
+
+fn distribution_point_urls(name: &DistributionPointName) -> Result<Vec<String>, Error> {
+    match name {
+        DistributionPointName::FullName(uris) => {
+            let uris: Result<Vec<String>, Error> = uris
+                .iter()
+                .map(|general_name| match general_name {
+                    GeneralName::UniformResourceIdentifier(s) => Ok(s.to_string()),
+                    _ => Err(Error::DistributionPointMalformed(
+                        "distribution point name something other than URI",
+                    )),
+                })
+                .collect();
+
+            uris
+        }
+        DistributionPointName::NameRelativeToCRLIssuer(_) => Err(
+            Error::DistributionPointMalformed("contained relative to issuer name"),
+        ),
+    }
+}
+
+async fn fetch_crl(url: &str) -> Result<Vec<u8>, Error> {
+    let bytes = reqwest::get(url)
```

Comment:
It would be good to create a `CrlVerifier` struct to avoid re-creating the reqwest client every time

Reply:
I'm don't believe that matches our data access pattern.  AFAIK, there isn't really a case where we need to batch many if even multiple requests for CRLs.